Encrypt init ns and proxy support

jrfastab opened the pull request:
Previously all traffic from Pod to Pod under Cilium mgmt was being encrypted but we omitted encrypting traffic from the init namespace and also traffic from "special" endpoints such as the health agent and proxy which do not run bpf_lxc.c BPF programs on egress. This patch resolves these and support to encrypt traffic that is routed through cilium_host but does not pass through the bpf_lxc BPF program.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6861)
<!-- Reviewable:end -->

jrfastab committed:
cilium: ipsec, host and proxy BPF support

Currently, with ipsec we do not encrypt traffic that is generated from the
host. This is traffic generated from the proxy, the health check agent
(which doesn't have an attached bpf_lxc), and any other traffic generated
through cilium_host interfaces.

This patch adds ipsec support to the BPF datapath for bpf_netdev. To do
this we use the cb to pass hints between bpf_netdev and a new program
bpf_ipsec that runs on the cilium_net ingress path. This program then
sets the mark/cb[4] values so that the traffic is encrypted by the xfrm
layer and so that when bpf_netdev receives the encrypted traffic the tunnel
id and secctx are known. This allows bpf_netdev to redirect the skb to the
tunnel device without further processing. In the future it would be nice to
avoid having to run the helper bpf program (bpf_ipsec) by doing the encrypt
operation directly from BPF, but this requires additional BPF kernel
helpers.

The flow is the following,
 1. skb routed to egress cilium_host causing bpf_netdev to run
 2. bpf_netdev identifies this as traffic to send out tunnel device
 3. if ipsec is enable set cb[0] and cb[4] with secctx and tunnel id
    if !ipsec redirect to tunnel device.
 4. skb needs ipsec, sent through veth pair causing ingress program of
    cilium_net (bpf_ipsec) to run.
 5. bpf_ipsec looks for hints in cb[0] and cb[4] then sets mark.
 6. xfrm layer matches mark and encrypts traffic
 7. route rule sends encrypted traffic to cilium_host causing bpf_netdev
    program to run.
 8. bpf_netdev program checks mark and because its set to encrypted
    extracts secctx and tunnel id then does redirect immediately to
    tunnel device.
 9. skb is encapsulated in tunnel

In the process this removes the last non-ipsec supported encap helper so we
can drop the ipsec postfix and encode it in the normal one.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

joestringer commented on bpf/init.sh:
```diff
@@ -435,6 +436,12 @@ if [ "$MODE" == "ipvlan" ]; then
 
+if [ "$MODE" = "vxlan" -o "$MODE" = "geneve" ]; then
+	if [ "$IPSEC" == "true" ]; then
+	bpf_load $HOST_DEV2 "$OPTS" "ingress" bpf_ipsec.c bpf_ipsec.o from-netdev $CALLS_MAP
```
nit: indentation

jrfastab replied:
sure easy to fix thanks.

joestringer commented on bpf/bpf_netdev.c:
```diff
@@ -241,8 +240,17 @@ static inline int handle_ipv6(struct __sk_buff *skb, __u32 src_identity)
+		 * but with encrypt marks.
+		 */
+		if (ret == IPSEC_ENDPOINT)
+			return redirect(HOST_IFINDEX, BPF_F_INGRESS);
```
This corresponds to the following line from the commit message right?

> skb needs ipsec, sent through veth pair causing ingress program of cilium_net (bpf_ipsec) to run.

The main point here being that we redirect to ingress on that device so that we don't lose the existing `skb->cb[]` fields, and hence everything will work OK.

jrfastab replied:
yeah that corresponds to the comment. But actually I'm going to convert that to 'return TC_ACT_OK' actually I already did it for ipv4.

The theory is we let the veth (cilium_host) pair forward it as normal, this calls dev_skb_forward into cilium_net. Here cb[] is in tact from above. Then bpf_ipsec can translate cb[] into mark values. Its really just a trick to revert clearing of the mark field done by dev_skb_forward between the veth pairs.

joestringer replied:
Ah, I kept thinking that it would be cleared when you cross the veth pair, but I suppose that's primarily the case when you're also crossing a netns boundary, not so much when both ends of the veth are in the init netns.

jrfastab replied:
The packet is "scrubbed" when you cross the veth pair so that mark and such are cleared but the cb[] is not so we can use like above. I don't believe cb[] is cleared upon crossing netns boundaries either which would be handled by the dev_skb_forward calling the scrub routine same as is done with veth pair in same netns.

My recollection is users of cb[] are responsible for clearing it before use. Xfrm code will do a memset before using cb[] for example. To avoid leaking data through cb[] users would want to clear any sensitive data before passing to next subsystem though.

joestringer commented on bpf/init.sh:
```diff
@@ -435,6 +436,12 @@ if [ "$MODE" == "ipvlan" ]; then
 fi
 bpf_load $HOST_DEV1 "$OPTS" "egress" bpf_netdev.c bpf_host.o from-netdev $CALLS_MAP
 
+if [ "$MODE" = "vxlan" -o "$MODE" = "geneve" ]; then
```
I forget, is direct-routing a follow-up PR later on? Do we need a similar trick there too?

(Less certain about if the traffic is from local host to a remote node's pod IP, but surely for post-proxy traffic it needs to go here)

jrfastab replied:
The direct routing is up next. The trick appears to be the exact same ;) But I'm not done testing it and didn't want to get into a spot where it broke something if enabled. That said at the moment the golang side doesn't have any stops if a user enables ipsec and direct mode. Currently its a don't do that. The direct routing patch should be posted shortly probably by Monday I hope.

joestringer reviewed (approved):
I marked this as needing backport to 1.4 because I think it's very surprising to people if they enable L7 visibility rules and suddenly their traffic is not encrypted. The alternative would be to document this in 1.4 as a deficiency in the implementation. (I guess really I'm just saying, whenever we have all we'll ship in the 1.4 tech preview of this feature, let's make sure the docs are clear about what's there and not).

I think just about all my comments below are just confirming my understanding, LGTM.

jrfastab committed:
cilium: ipsec, add node policies when node IPs not already included

If the node IPv{4|6} is not in the endpoint IP allocation range then we do not
currently have a policy in place for it so the traffic is sent on the wire
unencrypted. This patch adds the additional policy if required.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

joestringer commented on pkg/datapath/linux/linux_defaults/linux_defaults.go:
```diff
@@ -35,4 +35,7 @@ const (
 	TunnelDeviceName = "cilium_vxlan"
+
+	// IPSec SPI value for node rules
+	IPSecNodeSPI = 2
```
Is there anything magic about this SPI, particular reason `2` is chosen?

jrfastab replied:
Nothing magic about 2 it is needed to avoid conflicting with the state (src,dst,spi) assigned to the endpoint range and the other state matching node traffic.

Given two xfrm states S_1: (src_1, dst_1, spi_1)  and S_2 (src_2, dst_2, spi_2) the kernel considers the states S_1, S_2 to be colliding if dst_1 == dst_2 and spi_1 == spi_2. But we need to have a states to match outgoing traffic from both endpoints srcs and node ip srcs to another nodes dst ip. So we get these states S_1: (endpoint_ip_range, remote_ip, spi) and S_2: (node_ip, remote_ip, spi) so it collides with naive implementation with single spi*=1. Simple solution is to just use a special node spi here hence above magic number.

jrfastab replied:
And to make this more explicit I'll add a IPSecEndpointSPI = 1 constant and avoid the if clause and passing '0' as a magic default case through. I think it makes it a bit nicer.
